tools/functional-tester/etcd-runner/command/global.go

```diff
@@ -61,41 +56,41 @@ func newClient(eps []string, timeout time.Duration) *clientv3.Client {
 	return c
 }
 
-func doRounds(rcs []roundClient, rounds int) {
-	var mu sync.Mutex
-	var wg sync.WaitGroup
+func doRounds(rcs []roundClient, rounds int, requests int) {
+	var (
+		wg sync.WaitGroup
+	)
 
 	wg.Add(len(rcs))
 	finished := make(chan struct{})
+	limiter := rate.NewLimiter(rate.Limit(reqRate), reqRate)
 	for i := range rcs {
 		go func(rc *roundClient) {
 			defer wg.Done()
-			for rc.progress < rounds {
+			for rc.progress < rounds || rounds <= 0 {
+				if err := limiter.WaitN(context.Background(), requests/len(rcs)); err != nil {
+					log.Panicf("rate limiter error %v", err)
+				}
+
 				for rc.acquire() != nil { /* spin */
 				}
 
-				mu.Lock()
```

Comment:
It's not reasoning about lock discipline, it's enforcing one and the code assumed it would be guaranteed exclusion by that contract. I just don't understand why it's worth bothering changing this if it's causing this much trouble?